doc/PROTOCOL-SEMANTICS.md

```diff
@@ -0,0 +1,221 @@
+particular that there is no provision for the server application to not know
+whether the client application received a unary Call's response or a streaming
+Call's Status.
+
+Unary Calls may be optimized to be half-duplex and treat each stream as a single
+communication unit. That is, on the client a unary Call may be delayed from
+being sent until the Half Close is ready to be sent and on the server the
+response may be delayed until the Trailers is ready.
+
+Unary Calls that terminate with a Status Code of OK must contain a response
+message. Unary Calls that terminate with a Status Code other than OK do not need
+a response message, and at the implementation's discretion the response message
+may be discarded if present.
+
+## Status
+
+A Status contains a "code" and a "description". The Status Description is a
+human-readable Unicode string for developer debugging. The Status Code is a
+value from a pre-defined list of such codes. While Status Code is best
+communicated to users by its name, it commonly is treated as an integer
+internally, and so each code has a numeric value.
+
+The valid Status Codes are:
+
+| Num | Name                |
+| --- | ------------------- |
+| 0   | OK                  |
+| 1   | CANCELLED           |
+| 2   | UNKNOWN             |
+| 3   | INVALID_ARGUMENT    |
+| 4   | DEADLINE_EXCEEDED   |
+| 5   | NOT_FOUND           |
+| 6   | ALREADY_EXISTS      |
+| 7   | PERMISSION_DENIED   |
+| 8   | RESOURCE_EXHAUSTED  |
+| 9   | FAILED_PRECONDITION |
+| 10  | ABORTED             |
+| 11  | OUT_OF_RANGE        |
+| 12  | UNIMPLEMENTED       |
+| 13  | INTERNAL            |
+| 14  | UNAVAILABLE         |
+| 15  | DATA_LOSS           |
+| 16  | UNAUTHENTICATED     |
+
+Using the OK Status Code for a Call may only be decided by server applications.
+Library implementations must not "fabricate" an OK Status Code; it may only
+communicate an OK Status Code that was provided to it. While there may be
+additional restrictions on Status Code usage like those detailed in
+[statuscodes.md](statuscodes.md), those restrictions are more to provide a
+cohesive experience instead of a core, fundamental requirement.
+
+## Metadata
+
+Metadata has keys with associated values. Each key can have multiple values.
+Keys are unordered, but values for a key are ordered. Keys are case insensitive,
+and commonly canonicalized by making lower case. APIs are permitted to require a
+canonical representation and that representation may be different than the
+"lower case" representation mentioned here. A key can be for ASCII or binary
+values. If a key is for binary values, its name must be suffixed with "-bin".
+Otherwise it is for ASCII values.
+
+ASCII values are discouraged from having leading or trailing whitespace. If such
+a value contains leading or trailing whitespace, the whitespace may be stripped.
+Multiple ASCII values for the same key may be joined together with "," (a comma)
```

Comment:
Is this really part of the semantic protocol?  I don't think I was aware of this.  I suspect we're not doing anything like this in C-core.